pkg/cmd/flags.go

```diff
@@ -0,0 +1,172 @@
+// field is because 99% of the use cases where --version is defined the flag
+// will be required, and so we cater for the common case. Meaning only those
+// subcommands that have --version as optional will need to set that field.
+type ServiceVersionFlagOpts struct {
+	Dst      *string
+	Optional bool
+	Action   kingpin.Action
+}
+
+// NewServiceVersionFlag defines a --version flag that accepts multiple values
+// such as 'latest', 'active' and numerical values which are then converted
+// into the appropriate service version.
+func (b Base) NewServiceVersionFlag(opts ServiceVersionFlagOpts, args ...string) {
+	clause := b.CmdClause.Flag("version", "Number of service version, 'latest', 'active' or 'editable'")
+	if !opts.Optional {
+		clause = clause.Required()
+	} else {
+		clause = clause.Action(opts.Action)
+	}
+	clause.StringVar(opts.Dst)
+}
+
+// NewAutoCloneFlag defines a --autoclone flag that will cause a clone of the
+// identified service version if its found to be active or locked.
+func (b Base) NewAutoCloneFlag(action kingpin.Action, dst *bool) {
+	b.CmdClause.Flag("autoclone", "Automatically clone the identified service version if it's 'locked' or 'active'").Action(action).BoolVar(dst)
+}
+
+// OptionalServiceVersion represents a Fastly service version.
+type OptionalServiceVersion struct {
+	OptionalString
+}
+
+// Parse returns a service version based on the given user input.
+func (sv *OptionalServiceVersion) Parse(sid string, c api.Interface) (*fastly.Version, error) {
+	vs, err := c.ListVersions(&fastly.ListVersionsInput{
+		ServiceID: sid,
+	})
+	if err != nil || len(vs) == 0 {
+		return nil, fmt.Errorf("error listing service versions: %w", err)
+	}
+
+	// Sort versions into descending order.
+	sort.Slice(vs, func(i, j int) bool {
+		return vs[i].Number > vs[j].Number
+	})
+
+	var v *fastly.Version
+
+	switch strings.ToLower(sv.Value) {
+	case "latest":
+		return vs[0], nil
+	case "active":
+		v, err = getLatestActiveVersion(vs)
+	case "editable":
+		v, err = getLatestEditableVersion(vs)
+	case "":
+		v, err = getLatestEditableVersion(vs)
+	default:
+		v, err = getSpecifiedVersion(vs, sv.Value)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return v, nil
+}
+
+// OptionalAutoClone defines a method set for abstracting the logic required to
+// identify if a given service version needs to be cloned.
+type OptionalAutoClone struct {
+	OptionalBool
+}
+
+// Parse returns a service version.
+//
+// The returned version is either the same as the input argument `v` or it's a
+// cloned version if the input argument was either active or locked.
+func (ac *OptionalAutoClone) Parse(v *fastly.Version, sid string, c api.Interface) (*fastly.Version, error) {
+	// if user didn't provide --autoclone flag
+	if !ac.Value && (v.Active || v.Locked) {
+		return nil, fmt.Errorf("service version %d is not editable", v.Number)
+	}
+	if ac.Value && (v.Active || v.Locked) {
+		version, err := c.CloneVersion(&fastly.CloneVersionInput{
```

Comment:
Does this produce any output?  If not, I'd consider dropping in here a line of output that says:

> Service version {a} is not editable, so it was automatically cloned because `--autoclone` is enabled.  Now operating on version {b}...